src/ota_stm32f.c

```diff
@@ -0,0 +1,228 @@
+      if ((p - base) % (16 * 1024 * flash_config) == 0) return true;
+    } else if (p == base + 16 * 1024 * 4 * flash_config) {
+      return true;
+    } else if ((p - base) % (128 * 1024 * flash_config) == 0) return true;
+  }
+  return false;
+}
+
+MG_IRAM static int flash_sector(volatile uint32_t *addr) {
+  char *base = (char *) s_mg_flash_stm32f.start;
+  char *end = base + s_mg_flash_stm32f.size;
+  bool addr_in_bank_2 = false;
+  if (flash_dual_bank() > 0 && addr >= (uint32_t *) (base + (end - base) / 2)) {
+    addr = (uint32_t*)((uint32_t)addr - (end - base) / 2);
+    addr_in_bank_2 = true;
+  }
+  volatile char *p = (char *) addr;
+  uint32_t flash_config = FLASH_CONFIG_16_64_128;
+  if (STM_DEV_ID >= 0x449) {
+    flash_config = FLASH_CONFIG_32_128_256;
+  }
+  int sector = -1;
+  if (p >= base && p < end) {
+    if (p < base + 16 * 1024 * 4 * flash_config) {
+      sector =  (p - base) / (16 * 1024 * flash_config);
+    } else if (p >= base + 64 * 1024 * flash_config &&
+              p < base + 128 * 1024 * flash_config) {
+      sector = 4;
+    } else {
+      sector = (p - base) / (128 * 1024 * flash_config) + 4;
+    }
+  }
+  if (sector == -1) return -1;
+  if (addr_in_bank_2) sector += 12; // a bank has 12 sectors
+  return sector;
+}
+
+MG_IRAM static bool flash_is_err() {
+  return MG_REG(FLASH_BASE + FLASH_SR) & ((MG_BIT(7) - 1) << 1);
+}
+
+MG_IRAM static void flash_wait() {
+  while (MG_REG(FLASH_BASE + FLASH_SR) & (MG_BIT(16))) (void) 0;
+}
+
+MG_IRAM static void flash_clear_err() {
+  flash_wait(); // Wait until ready
+  MG_REG(FLASH_BASE + FLASH_SR) = 0xf2;  // Clear all errors
+}
+
+MG_IRAM bool mg_stm32f_erase(void *addr) {
+  bool ok = false;
+  if (flash_page_start(addr) == false) {
+    MG_ERROR(("%p is not on a sector boundary", addr));
+  } else {
+    int sector = flash_sector(addr);
+    if (sector < 0) return false;
+    uint32_t sector_reg = sector;
+    if (flash_dual_bank() > 0 && sector >= 12) {
+      // 3.9.8 Flash control register (FLASH_CR) for F42xxx and F43xxx
+      // BITS[7:3]
+      sector_reg -= 12;
+      sector_reg |= MG_BIT(4);
+    }
+    flash_unlock();
+    flash_wait();
+    uint32_t cr = MG_BIT(1); // SER
+    cr |= MG_BIT(16); // STRT
+    cr |= (sector_reg & 31) << 3; // sector
+    MG_REG(FLASH_BASE + FLASH_CR) = cr;
+    ok = !flash_is_err();
+    MG_DEBUG(("Erase sector %lu @ %p %s. CR %#lx SR %#lx", sector, addr,
+              ok ? "ok" : "fail", MG_REG(FLASH_BASE + FLASH_CR),
+              MG_REG(FLASH_BASE + FLASH_SR)));
+    // After we have erased the sector, set CR flags for programming
+    // 2 << 8 is word write parallelism, bit(0) is PG. RM0385, section 3.7.5
+    MG_REG(FLASH_BASE + FLASH_CR) = MG_BIT(0) | (2 << 8);
+    flash_clear_err();
+  }
+  return ok;
+}
+
+MG_IRAM bool mg_stm32f_swap(void) {
+  return true;
+}
+
+static bool s_flash_irq_disabled;
+
+MG_IRAM bool mg_stm32f_write(void *addr, const void *buf, size_t len) {
```

Comment:
static

Reply:
I'd rather not mark this with static.

The reason is that the compiler generates an extra symbol when we declare it with static, because this RAM function is called from another function from flash. 

```
1121: 0800e809     8 FUNC    LOCAL  DEFAULT    2 __mg_stm32f_writ[...]
```
Notice that the extra symbol __mg_stm32f_writ is linked in flash. During the firmware override, once we erase the corresponding sector, I think we cannot return from mg_stm32f_write anymore. Firmware override works, but I noticed that the MCU does not reset, I have to reset it manually to boot, so I assume that in single_bank_swap function, the instruction for resetting is never reached, because we are not returning from mg_stm32f_write.

Unless you know some compiler arguments that could prevent creating that extra symbol, we could leave it without static, because it works this way.

Reply:
I noticed my compiler generates a veneer for _iram_ functions, didn't do a deep dive as code was working for me with MCXN947 and for you with H723/35. Those have all marked functions as static.
mg_stm32f_write() must and does return.
The only function that does not return is single_bank_swap(), it has just changed any possible return point.
The first thing I notice is that you don't do as I did, I performed all calculations on flash code inside mg_ota_end() and then called a simplest _iram_ function with lots of parameters. We can't have any stdlib calls in single_bank_swap() (I left a nice allegoric comment there). What if the compiler decides to call stdlib for the /2 ? Yeah, it happens before the actual erase and it should work, but it is not working for some reason...
I have F746 and F429 boards, so I can try this.
We should disassemble what happens from when mg_ota_end() calls single_bank_swap() and trace to the point where we reset. That's what ICEs are for.
Maybe your compiler is different, though that should have made the H723 to work different.
Let's simplify that function, make it static, and see what happens and find the reason for it.

Reply:
Upon a closer inspection, I found that the compiler does not actually generate the single_bank_swap function, instead it inlines it. So, the _mg_stm32f_write veneer is called from mg_ota_end and after that it executes the reset instruction again in mg_ota_end.

Shall we force the compiler to avoid inlining for single_swap_bank?

```
__attribute__((noinline)) MG_IRAM static void single_bank_swap(char *p1, char *p2, size_t size)
```

Reply:
Yes, that's exactly was I looking at yesterday night with the Pico-SDK.
Darned compilers that think they know more than programmers... In my case, I saw single_bank_swap() was generated, not inlined.
single_bank_swap() is a short function, but I think we should apply that to write and erase too. I don't care if it's optimal to inline erase into write, I see two separate functions and I want two separate functions... otherwise my reasoning is bogus.
Maybe the rest of the calls can be safely inlined into these two, so I'm not into adding that inside MG_IRAM, BUT, we should have MG_NOINLINE and apply it to erase, write, and single_bank_swap. The swap function should be analyzed case by case... and we need to state this in a comment.
I'll take care of the others, let's make this one work.

Reply:
Probably in my case the compiler generated single_bank_swap() because it is a loop with locals (?) ... Anyway, we can't rely on heuristics that might change from version to version